File: runtime-src/Monk/Runtime/Child.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}

-- | Portable byte transport for already materialized Fish children. No source
-- language expressions are accepted or interpreted by this module.
module Monk.Runtime.Child (dispatchChild, dispatchChildInSession) where

import Control.Concurrent (myThreadId, throwTo)
import Control.Exception (IOException, bracket, bracketOnError, catch, onException)
import Control.Monad (filterM, forM, forM_, unless, void, when)
import Data.Bits (testBit)
import Data.ByteString qualified as B
import Data.ByteString.Char8 qualified as C
import Data.Map.Strict qualified as M
import GHC.Foreign qualified as Foreign
import GHC.IO.Encoding (getFileSystemEncoding)
import Monk.Runtime.Descriptors (duplicatePrivate, initialDescriptorOpen)
import Monk.Runtime.Session (dispatchSessionChild)
import Monk.Runtime.Spawn (initialSignalIgnored, spawnProcess)
import System.Directory (getTemporaryDirectory, removeDirectoryRecursive)
import System.Environment (getExecutablePath)
import System.Exit (ExitCode (..), exitWith)
import System.IO (Handle, hClose, hFlush, openBinaryTempFile, stdin, stdout)
import System.Posix.Env.ByteString (getEnvironment)
import System.Posix.IO (closeFd, createPipe, fdToHandle)
import System.Posix.Process (ProcessStatus (..), getProcessStatus)
import System.Posix.Signals (Handler (Catch), installHandler, sigHUP, sigINT, sigKILL, sigPIPE, sigQUIT, sigTERM, signalProcess)
import System.Posix.Temp (mkdtemp)
import System.Posix.Types (Fd (..), ProcessID)
import Text.Read (readMaybe)

-- | The first four NUL-terminated frames are diagnostic bytes, descriptor mask, Fish script,
-- and SHLVL. The remaining bytes are the child's NUL-framed scalar/argv state.
dispatchChild :: Bool -> IO ()
dispatchChild = dispatchChildWith False

dispatchChildInSession :: Bool -> IO ()
dispatchChildInSession = dispatchChildWith True

dispatchChildWith :: Bool -> Bool -> IO ()
dispatchChildWith supervised capture = run `catch` failure
  where
    run = do
      (code, output) <- withOwnerSignals $ do
        bytes <- B.hGetContents stdin
        (warning, descriptorMask, script, level, state) <- either (ioError . userError) pure (decode bytes)
        originalInput <- initialDescriptorOpen 3
        when (testBit descriptorMask 0 && not originalInput) (ioError (userError "missing original stdin descriptor"))
        if supervised
          then dispatchSessionChild capture warning descriptorMask script level state
          else do
            environment <- getEnvironment
            let childEnvironment = ("SHLVL", C.pack (show level)) : filter ((/= "SHLVL") . fst) environment
            directory <- getTemporaryDirectory
            bracket (mkdtemp (directory <> "/monk-child-XXXXXX")) removeDirectoryRecursive $ \workspace -> do
              scriptPath <- transportFile workspace (restoreClosedStreams capture descriptorMask script)
              statePath <- transportFile workspace state
              if capture
                then bracket capturePipe (\(r, w, _) -> closeHandleQuiet r >> closeHandleQuiet w) $ \(reader, writer, writeFd) ->
                  bracketOnError (spawnChild supervised descriptorMask scriptPath statePath childEnvironment (Just writeFd)) terminateChild $ \pid -> do
                    hClose writer
                    output <- drainCapture (testBit descriptorMask 2) warning reader
                    code <- waitChild pid
                    pure (code, B.dropWhileEnd (== 10) output)
                else bracketOnError (spawnChild supervised descriptorMask scriptPath statePath childEnvironment Nothing) terminateChild $ \pid -> do
                  code <- waitChild pid
                  pure (code, B.empty)
      -- Publish only after owned transport resources have been released. A
      -- reader closing the packet pipe must not strand private files.
      if capture then packet "ok" code output else exitWith (if code == 0 then ExitSuccess else ExitFailure code)
    failure (_ :: IOException) =
      if capture
        then packet "error" 125 "child-transport-failure" `catch` (\(_ :: IOException) -> exitWith (ExitFailure 125))
        else writeDiagnostic "monk: child transport failed\n" >> exitWith (ExitFailure 125)

decode :: B.ByteString -> Either String (B.ByteString, Int, B.ByteString, Integer, B.ByteString)
decode input = do
  (warning, rest) <- frame input
  (maskBytes, payload) <- frame rest
  descriptorMask <- maybe (Left "invalid descriptor mask") Right (readMaybe (C.unpack maskBytes))
  unless (descriptorMask >= 0 && descriptorMask <= 7) (Left "invalid descriptor mask")
  (script, rest') <- frame payload
  (levelBytes, state) <- frame rest'
  level <- maybe (Left "invalid SHLVL") Right (readMaybe (C.unpack levelBytes))
  unless (B.null state || B.last state == 0) (Left "unterminated state")
  pure (warning, descriptorMask, script, level, state)
  where
    frame bytes = case B.break (== 0) bytes of
      (part, tailBytes) | not (B.null tailBytes) -> Right (part, B.tail tailBytes)
      _ -> Left "missing frame"

-- Files live inside a mode-0700 mkdtemp directory, and openBinaryTempFile
-- creates each mode-0600 file. The owning bracket removes the directory after
-- wait, including exceptions. Fish opens its own independent offset-zero file
-- descriptions; neither /proc nor /dev/fd reopening is part of the ABI.
transportFile :: FilePath -> B.ByteString -> IO B.ByteString
transportFile directory bytes = do
  path <- bracket (openBinaryTempFile directory "transport") (closeHandleQuiet . snd) $ \(path, handle) -> do
    B.hPut handle bytes
    hFlush handle
    pure path
  -- FilePath can contain surrogate escapes for non-UTF-8 native path bytes.
  encoding <- getFileSystemEncoding
  Foreign.withCString encoding path B.packCString

capturePipe :: IO (Handle, Handle, Fd)
capturePipe = bracket createPipe (\(r, w) -> closeQuiet r >> closeQuiet w) $ \(r, w) -> do
  r' <- duplicatePrivate r
  w' <- duplicatePrivate w `onException` closeFd r'
  reader <- fdToHandle r' `onException` (closeFd r' >> closeFd w')
  writer <- fdToHandle w' `onException` (hClose reader >> closeFd w')
  pure (reader, writer, w')

-- Prepare every descriptor and argv in the parent. The spawn implementation
-- performs only native file/signal actions before exec, never Haskell after fork.
spawnChild :: Bool -> Int -> B.ByteString -> B.ByteString -> [(B.ByteString, B.ByteString)] -> Maybe Fd -> IO ProcessID
spawnChild supervised descriptorMask scriptPath statePath environment output = do
  let input = [(0, Fd 3) | testBit descriptorMask 0]
      standardOutput = maybe [(1, Fd 1) | testBit descriptorMask 1] (\fd -> [(1, fd)]) output
      diagnostic = [(2, Fd 2) | testBit descriptorMask 2]
      descriptors = M.fromList (input <> standardOutput <> diagnostic)
  if supervised
    then do
      runtimePath <- getExecutablePath
      encoding <- getFileSystemEncoding
      runtime <- Foreign.withCString encoding runtimePath B.packCString
      spawnProcess descriptors Nothing environment runtime ["--abi", "2", "session-run", scriptPath, statePath]
    else spawnProcess descriptors Nothing environment "fish" ["--no-config", scriptPath, statePath]

-- The owner receives cleanup exceptions rather than abandoning transport
-- resources on termination. Its direct child is reaped before files disappear.
withOwnerSignals :: IO a -> IO a
withOwnerSignals action = do
  owner <- myThreadId
  let install signal = do
        previous <- installHandler signal (Catch (throwTo owner (ExitFailure (128 + fromIntegral signal)))) Nothing
        pure (signal, previous)
      restore handlers = forM_ handlers $ \(signal, previous) -> void (installHandler signal previous Nothing)
  signals <- filterM (fmap not . initialSignalIgnored) [sigHUP, sigINT, sigQUIT, sigTERM, sigPIPE]
  bracket (forM signals install) restore (const action)

terminateChild :: ProcessID -> IO ()
terminateChild pid = do
  signalProcess sigKILL pid `catch` ignore
  void (waitChild pid) `catch` ignore

-- Fish repairs absent standard streams at startup. Reapply the original
-- closures inside its script so builtin writes and external children observe
-- the caller's descriptor contract rather than silently writing to /dev/null.
restoreClosedStreams :: Bool -> Int -> B.ByteString -> B.ByteString
restoreClosedStreams capture descriptorMask script =
  "begin\n" <> script <> "\nend" <> B.concat [" " <> C.pack (show fd) <> ">&-" | fd <- [0 :: Int .. 2], not (testBit descriptorMask fd), not (capture && fd == 1)] <> "\n"

waitChild :: ProcessID -> IO Int
waitChild pid = do
  status <- getProcessStatus True False pid
  case status of
    Just (Exited ExitSuccess) -> pure 0
    Just (Exited (ExitFailure code)) -> pure code
    Just (Terminated signal _) -> pure (128 + fromIntegral signal)
    _ -> pure 125

-- Drain in bounded reads so the first NUL diagnostic is emitted while the
-- child is running. Delay only the final packet, whose trailing newlines cannot
-- be decided until EOF. Raw non-NUL bytes never cross a text decoder.
drainCapture :: Bool -> B.ByteString -> Handle -> IO B.ByteString
drainCapture stderrOpen warning handle = loop False []
  where
    loop warned chunks = do
      chunk <- B.hGetSome handle 65536
      if B.null chunk
        then pure (B.concat (reverse chunks))
        else do
          let nul = B.elem 0 chunk
          when (nul && not warned && stderrOpen) (writeDiagnostic warning)
          loop (warned || nul) (B.filter (/= 0) chunk : chunks)

packet :: B.ByteString -> Int -> B.ByteString -> IO ()
packet tag code bytes = B.hPut stdout (tag <> "\0" <> C.pack (show code) <> "\0" <> bytes <> "\0") >> hFlush stdout

writeDiagnostic :: B.ByteString -> IO ()
writeDiagnostic bytes = (do fd <- duplicatePrivate (Fd 2); bracket (fdToHandle fd) hClose (\handle -> B.hPut handle bytes >> hFlush handle)) `catch` ignore

closeHandleQuiet :: Handle -> IO ()
closeHandleQuiet handle = hClose handle `catch` ignore

closeQuiet :: Fd -> IO ()
closeQuiet fd = closeFd fd `catch` ignore

ignore :: IOException -> IO ()
ignore _ = pure ()

File: runtime/src/child.rs
//! Framed transport for materialized Fish children and supervised child callback.
use crate::native::Workspace;
use crate::{
    native,
    types::{DescriptorMask, SourceFd},
};
use std::{
    io,
    os::fd::{AsFd, BorrowedFd},
};

pub struct Request {
    pub warning: Vec<u8>,
    pub mask: DescriptorMask,
    pub script: Vec<u8>,
    pub level: Vec<u8>,
    pub state: Vec<u8>,
}
pub type SessionCallback = fn(bool, Request) -> io::Result<(i32, Vec<u8>)>;
pub fn decode(bytes: &[u8]) -> io::Result<Request> {
    let bad = || io::Error::new(io::ErrorKind::InvalidInput, "invalid child transport");
    let mut rest = bytes;
    let mut frame = || -> io::Result<Vec<u8>> {
        let end = rest.iter().position(|b| *b == 0).ok_or_else(bad)?;
        let value = rest[..end].to_vec();
        rest = &rest[end + 1..];
        Ok(value)
    };
    let warning = frame()?;
    let mask = frame()?;
    let script = frame()?;
    let level = frame()?;
    let mask = std::str::from_utf8(&mask)
        .ok()
        .and_then(|s| s.parse::<u8>().ok())
        .ok_or_else(bad)?;
    let mask = DescriptorMask::new(mask)?;
    let mut digits = level.as_slice();
    if digits.first() == Some(&b'-') || digits.first() == Some(&b'+') {
        digits = &digits[1..];
    }
    if digits.is_empty() || !digits.iter().all(u8::is_ascii_digit) {
        return Err(bad());
    }
    if !rest.is_empty() && rest.last() != Some(&0) {
        return Err(bad());
    }
    Ok(Request {
        warning,
        mask,
        script,
        level,
        state: rest.to_vec(),
    })
}
pub fn dispatch(capture: bool, supervised: Option<SessionCallback>) -> io::Result<i32> {
    let result = (|| {
        let _signals = native::SignalGuard::install()?;
        let input = native::inherited_fd(0)?;
        let bytes = read_all(input.as_fd())?;
        let request = decode(&bytes)?;
        if request.mask.contains(SourceFd::new(0)?) && !native::initial_descriptor_open(3) {
            return Err(io::Error::new(
                io::ErrorKind::InvalidInput,
                "missing original stdin descriptor",
            ));
        }
        if let Some(callback) = supervised {
            callback(capture, request)
        } else {
            run(capture, request)
        }
    })();
    if let Some(signal) = native::pending_signal() {
        return Ok(128 + signal);
    }
    match result {
        Ok((code, output)) => {
            if capture {
                packet(b"ok", code, &output)?;
                Ok(0)
            } else {
                Ok(code)
            }
        }
        Err(_) => {
            if capture {
                match packet(b"error", 125, b"child-transport-failure") {
                    Ok(()) => Ok(0),
                    Err(_) => Ok(125),
                }
            } else {
                diagnostic(b"monk: child transport failed\n");
                Ok(125)
            }
        }
    }
}
fn run(capture: bool, request: Request) -> io::Result<(i32, Vec<u8>)> {
    let workspace = Workspace::new(b"monk-child-")?;
    let script = workspace.file(&restore_closed_streams(
        capture,
        request.mask,
        &request.script,
    ))?;
    let state = workspace.file(&request.state)?;
    let mut environment = native::environment();
    environment.retain(|(key, _)| key != b"SHLVL");
    environment.insert(0, (b"SHLVL".to_vec(), normalize_integer(&request.level)));
    let mut streams = native::Streams::new();
    for fd in 0..3 {
        let source = SourceFd::new(fd)?;
        if request.mask.contains(source) {
            let inherited = if fd == 0 { 3 } else { fd };
            streams.insert(source, native::inherited_fd(inherited)?);
        }
    }
    let pipe = if capture {
        let (reader, writer) = native::private_pipe()?;
        streams.insert(SourceFd::new(1)?, writer);
        Some(reader)
    } else {
        None
    };
    let mut child = native::spawn(
        &streams,
        None,
        &environment,
        b"fish",
        &[b"--no-config".to_vec(), script, state],
        false,
    )?;
    drop(streams);
    let output_result = if let Some(reader) = pipe {
        drain_capture(
            reader.as_fd(),
            request.mask.contains(SourceFd::new(2)?),
            &request.warning,
        )
    } else {
        Ok(Vec::new())
    };
    let output = match output_result {
        Ok(output) => output,
        Err(error) => {
            let _ = child.terminate();
            return Err(error);
        }
    };
    let code = child.complete()?.outcome().code();
    Ok((code, output))
}

fn normalize_integer(value: &[u8]) -> Vec<u8> {
    let negative = value.first() == Some(&b'-');
    let digits = value
        .strip_prefix(b"-")
        .or_else(|| value.strip_prefix(b"+"))
        .unwrap_or(value);
    let start = digits
        .iter()
        .position(|b| *b != b'0')
        .unwrap_or(digits.len());
    if start == digits.len() {
        return b"0".to_vec();
    }
    if negative {
        [b"-".as_slice(), &digits[start..]].concat()
    } else {
        digits[start..].to_vec()
    }
}
pub fn restore_closed_streams(capture: bool, mask: DescriptorMask, script: &[u8]) -> Vec<u8> {
    let mut result = [b"begin\n".as_slice(), script, b"\nend"].concat();
    for fd in 0..3 {
        if mask.get() & (1 << fd) == 0 && !(capture && fd == 1) {
            result.extend_from_slice(format!(" {fd}>&-").as_bytes());
        }
    }
    result.push(b'\n');
    result
}
pub fn read_all(fd: BorrowedFd<'_>) -> io::Result<Vec<u8>> {
    crate::transport::read_all(fd)
}
fn drain_capture(fd: BorrowedFd<'_>, stderr_open: bool, warning: &[u8]) -> io::Result<Vec<u8>> {
    crate::capture::drain(
        fd,
        || {
            if native::pending_signal().is_some() {
                Err(io::ErrorKind::Interrupted.into())
            } else {
                Ok(())
            }
        },
        || {
            if stderr_open {
                diagnostic(warning);
            }
            Ok(())
        },
    )
}
fn packet(tag: &[u8], code: i32, bytes: &[u8]) -> io::Result<()> {
    native::write_all(
        native::inherited_fd(1)?.as_fd(),
        &[tag, b"\0", code.to_string().as_bytes(), b"\0", bytes, b"\0"].concat(),
    )
}
fn diagnostic(bytes: &[u8]) {
    if let Ok(fd) = native::inherited_fd(2) {
        let _ = native::write_all(fd.as_fd(), bytes);
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn child_frames_keep_raw_bytes() {
        let request = decode(
            b"warn\0"
                .iter()
                .chain(b"5\0\xff\0-0005\0a\0")
                .copied()
                .collect::<Vec<_>>()
                .as_slice(),
        )
        .unwrap();
        assert_eq!(request.script, b"\xff");
        assert_eq!(request.state, b"a\0");
        assert_eq!(normalize_integer(&request.level), b"-5");
    }
    #[test]
    fn child_rejects_bad_mask_and_unterminated_state() {
        assert!(
            decode(
                b"\0"
                    .iter()
                    .chain(b"8\0script\x001\0")
                    .copied()
                    .collect::<Vec<_>>()
                    .as_slice()
            )
            .is_err()
        );
        assert!(decode(b"w\x007\x00s\x001\x00bad").is_err());
    }
    #[test]
    fn capture_restores_only_original_closed_streams() {
        assert_eq!(
            restore_closed_streams(true, DescriptorMask::new(0).unwrap(), b"printf x"),
            b"begin\nprintf x\nend 0>&- 2>&-\n"
        );
    }
}
